opentelemetry-sdk/benches/log_exporter.rs

```diff
@@ -25,147 +31,232 @@ use opentelemetry_sdk::logs::LogRecord;
+    use super::*;
 
-pub trait LogExporterWithoutFuture: Send + Sync + Debug {
-    fn export(&mut self, batch: LogBatch<'_>);
-}
+    /// Async style using the `async_trait` crate
+    #[async_trait]
+    pub trait LogExporterAsyncTrait: Send + Sync + Debug {
+        async fn export(&mut self, batch: LogBatch<'_>);
+    }
 
-#[derive(Debug)]
-struct NoOpExporterWithFuture {}
+    #[derive(Debug)]
+    pub struct NoOpExporterAsyncTrait;
 
-#[async_trait]
-impl LogExporterWithFuture for NoOpExporterWithFuture {
-    async fn export(&mut self, _batch: LogBatch<'_>) {}
-}
+    #[async_trait]
+    impl LogExporterAsyncTrait for NoOpExporterAsyncTrait {
+        async fn export(&mut self, batch: LogBatch<'_>) {
+            // As of writing this benchmark, the `rust-version` is `1.65`, but
+            // this benchmark relies on native async trait support from `1.75`
+            // so we may as well use the `black_box` hint from `1.66`. Benchmark
+            // are expected to be executed with a recent Rust toolchain anyway.
+            #[allow(clippy::incompatible_msrv)]
+            black_box(batch);
+        }
+    }
 
-#[derive(Debug)]
-struct NoOpExporterWithoutFuture {}
-impl LogExporterWithoutFuture for NoOpExporterWithoutFuture {
-    fn export(&mut self, _batch: LogBatch<'_>) {}
-}
+    #[derive(Debug)]
+    pub struct ExportingProcessorAsyncTrait {
+        exporter: Mutex<NoOpExporterAsyncTrait>,
+    }
 
-#[derive(Debug)]
-struct ExportingProcessorWithFuture {
-    exporter: Mutex<NoOpExporterWithFuture>,
-}
+    impl ExportingProcessorAsyncTrait {
+        pub fn new(exporter: NoOpExporterAsyncTrait) -> Self {
+            Self {
+                exporter: Mutex::new(exporter),
+            }
+        }
+    }
+
+    impl LogProcessor for ExportingProcessorAsyncTrait {
+        fn emit(&self, record: &mut LogRecord, library: &InstrumentationLibrary) {
+            let mut exporter = self.exporter.lock().expect("lock error");
+            let logs = [(record as &LogRecord, library)];
+            futures_executor::block_on(exporter.export(LogBatch::new(&logs)));
+        }
 
-impl ExportingProcessorWithFuture {
-    fn new(exporter: NoOpExporterWithFuture) -> Self {
-        ExportingProcessorWithFuture {
-            exporter: Mutex::new(exporter),
+        fn force_flush(&self) -> LogResult<()> {
+            Ok(())
+        }
+
+        fn shutdown(&self) -> LogResult<()> {
+            Ok(())
         }
     }
 }
 
-impl LogProcessor for ExportingProcessorWithFuture {
-    fn emit(&self, record: &mut LogRecord, library: &InstrumentationLibrary) {
-        let mut exporter = self.exporter.lock().expect("lock error");
-        let logs = [(record as &LogRecord, library)];
-        futures_executor::block_on(exporter.export(LogBatch::new(&logs)));
+mod style_async_native {
+    use super::*;
+
+    /// Async style using Rustc support for async traits
+    pub trait LogExporterAsyncNative: Send + Sync + Debug {
```

Comment:
From the benchmarks, this looks to be close enough with sync counterpart. Wondering if there is a reason we can't try this in the Exporter traits? The <T> for output might make things complex...?
(I only added this benchmarks to see if there is actually some price to be paid for introducing async.)

@demurgos Do you have some bandwidth to modify existing exporter trait to use async native style? The existing benchmarks should show improvement with this change, based on what i observe in this PR.

(Still pretty new to Rust async, so I might be missing something obvious!)

Reply:
Hey :wave: 
I was also interested in trying to impl a version using native async. I'm pretty familiar with it, but the lib is pretty large. I'll try to send something, but I can't promise any timeline.